[dtensor] use tracing for metadata prop

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95655
* #95458
* #95457
* __->__ #95456

This PR uses tracing for metadata prop, so that we can get correct
shape/stride metadata without manual calculation by ourselves.

The follow up PR on this would be adopt tracing for the sharding
prop itself

Differential Revision: [D43643578](https://our.internmc.facebook.com/intern/diff/D43643578)

wanchaol committed:
[dtensor] use tracing for metadata prop

This PR uses tracing for metadata prop, so that we can get correct
shape/stride metadata without manual calculation by ourselves.

The follow up PR on this would be adopt tracing for the sharding
prop itself

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] use tracing for metadata prop"


This PR uses tracing for metadata prop, so that we can get correct
shape/stride metadata without manual calculation by ourselves.

The follow up PR on this would be adopt tracing for the sharding
prop itself

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] use tracing for metadata prop"


This PR uses tracing for metadata prop, so that we can get correct
shape/stride metadata without manual calculation by ourselves.

The follow up PR on this would be adopt tracing for the sharding
prop itself

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM! Is the reason we move metadata `ndim` and `shape` into tensor that we want to have `fx` trace them?

wanchaol committed:
Update on "[dtensor] use tracing for metadata prop"


This PR uses tracing for metadata prop, so that we can get correct
shape/stride metadata without manual calculation by ourselves.

The follow up PR on this would be adopt tracing for the sharding
prop itself

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43643578).
